### PR TITLE
3D group forward convolution solver

### PR DESCRIPTION
**Pull request by iq136boy**
Enable the 3d group forward convolution support in miopen using the composable kernel.

### CODE REVIEW

File: Dockerfile
Comment (averinevg) [inline Dockerfile]: Why do we need this line?

File: fin
Comment (averinevg) [inline fin:1]: Please use the most recent commit from MIFin ([74382a6](https://github.com/ROCmSoftwarePlatform/MIFin/commit/74382a6746d5aa0117f15d0a7cb779607caaa78d))

**Review by averinevg — CHANGES_REQUESTED** [review]

File: Dockerfile
Comment (iq136boy) [reply Dockerfile]: > Why do we need this line?

This is just for the access for my local /tmp directory. I will delete it. Thanks.

**Review by averinevg — DISMISSED** [review]
LGTM

File: requirements.txt
Comment (DrizztDoUrden) [inline requirements.txt]: Did something change in the line? I don't see anything.

**Review by DrizztDoUrden — DISMISSED** [review]
LGTM

File: requirements.txt
Comment (junliume) [reply requirements.txt]: @iq136boy @DrizztDoUrden An empty line is added to the end of this file, which I believe should not be needed.

**Review by junliume — DISMISSED** [review]

**Comment by junliume** [comment]
Ping @JehandadKhan  before merging PR

File: src/include/miopen/solver.hpp
Comment (atamazov) [inline src/include/miopen/solver.hpp]: [Recommendation][Quality] There is the same comment at line 4613. Add there:
```C++
    /// \anchor igemm_get_wti_magic_number
```
```suggestion
    /// \ref igemm_get_wti_magic_number
```
The same "ref" stuff is recommended for use at lines 4695 and 4779.

File: src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp]: [Correctness]
```suggestion
    if(problem.conv_problem.GetConv().attribute.deterministic)
```

File: src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp]: [R] This "elapsed_time" stuff is not necessary when the Solver enqueues one kernel.

File: src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp]: Ditto.

File: src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp]: [Performance] The most of these computations can (and should) be moved to the _Invoker Factory_ (outer lambda) or to the solver (i.e. can be computed right in GetSolution). Again, the library creates a dedicated instance of _Invoker_ for each specific problem. In the Invoker's context, the problem is **_constant_**. Therefore parsing its values and building parameters in the Invoker's context is a waste of CPU time.

**Review by atamazov — CHANGES_REQUESTED** [review]

**Comment by JehandadKhan** [comment]
@amberhassaan  and @iq136boy  Please address @atamazov review comments so we can merge this.

File: src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp
Comment (iq136boy) [reply src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp]: @atamazov  I will refactor the code as suggested. Thanks.

**Comment by iq136boy** [comment]
> @amberhassaan and @iq136boy Please address @atamazov review comments so we can merge this.

@JehandadKhan Got it. I will address the comments. Thanks.

**Comment by junliume** [comment]
Plan to merge 08/11, ping @atamazov  @JehandadKhan

**Review by averinevg — DISMISSED** [review]

File: src/include/miopen/solver.hpp
Comment (atamazov) [reply src/include/miopen/solver.hpp]: [Resolved]

File: src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp
Comment (atamazov) [reply src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp]: [Resolved]

File: src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp
Comment (atamazov) [reply src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp]: [Resolved]

File: src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp
Comment (atamazov) [reply src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp]: [Resolved]

File: src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp
Comment (atamazov) [reply src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp]: @iq136boy Original comment updated: added notice that computations can be moved even higher, into `GetSolution`.

Comment (atamazov) [reply src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp]: @iq136boy Unfortunately, the code basically remains the same. Let me clarify: moving the code into a function does not change anything (and in this case highly recommended to be reverted unless you want to call the function from two or more places in the code). `Invoker` is an inner lambda, and what we need to do is move computations out of it. AFAICS the most of excessive computations are performed in RunCKSolution.

Please refer at #2305 -- this PR is intended to resolve the same problem. Hope this helps.

**Comment by atamazov** [comment]
@junliume 
> Plan to merge 08/11, ping @atamazov @JehandadKhan

I am not going to block the merge (no correctness problems), but please be informed that actual performance of this solver may be worse than expected due to https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2286#discussion_r1286007620.

**Review by junliume — DISMISSED** [review]

**Comment by JehandadKhan** [comment]
@averinevg  Will update this PR to reflect the new problem description changes.

File: src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp
Comment (iq136boy) [reply src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp]: @atamazov Thanks for the [clarification](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2286#discussion_r1291779229)
I have a quick fix to move parts of the computation from invoker. For more refactoring work to improve performance will be addressed in a separate PR.

**Review by junliume — APPROVED** [review]
LGTM, and watching last round of CI

File: src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp:263]: I think that computation of `conv_ptr` should be moved to GetSolution or to InvokerFactory at least. See https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2305#discussion_r1290087188 for more info.

File: src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp:296]: [R] Get rid of this function as it is called only in one place.

File: src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp:307]: Move switch out of the Invoker, see https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2305#discussion_r1290090595.

**Review by atamazov — COMMENTED** [review]
Again, I am not going to block the merge, but these review comments must be resolved sooner or later.

**Comment by junliume** [comment]
@atamazov thanks for the reviews! :)
@iq136boy please create a post-merge issue and deal with the latest review opinions in the next PR. Thanks!

File: src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp
Comment (amberhassaan) [reply src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp:263]: Fyi: I was also trying to move `conv_ptr` computation out of the invoker lambda and it leads to compilation errors that can't be resolved without major changes. See https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2286#pullrequestreview-1578557778

File: src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp
Comment (amberhassaan) [reply src/solver/conv_hip_implicit_gemm_3d_grouped_fwd_xdlops.cpp:307]: Switch can be moved out. Please see 9034f3e in #2305

**Comment by iq136boy** [comment]
@junliume @atamazov Based on the discussion with Jehandad, Amber will take over the following refactoring work for this PR. And I will deal with the reviews in the new 3d backward data and backward weight solver that I am currently working on.

**Comment by atamazov** [comment]
@iq136boy 
> Based on the discussion with Jehandad, Amber will take over the following refactoring work for this PR.

Thanks for informing. Can you please clarify who will create a post merge issue?

/CC @amberhassaan @junliume @JehandadKhan

**Comment by junliume** [comment]
> @iq136boy
> 
> > Based on the discussion with Jehandad, Amber will take over the following refactoring work for this PR.
> 
> Thanks for informing. Can you please clarify who will create a post merge issue?
> 
> /CC @amberhassaan @junliume @JehandadKhan

@amberhassaan could you take on the post merge issue? recommendation:
1. First create a new issue referring to the review comments which are not resolved here. Label the urgency level.
2. Create a PR fixing the issue.

**Comment by amberhassaan** [comment]
@junliume : acknowledging. Will create the issue and follow up PR.

**Pull request merged**
